cpu/lpc2387/periph/uart.c

```diff
@@ -29,84 +30,251 @@
 #include "VIC.h"
 #include "periph/uart.h"
 
-/* for now, we only support one UART device... */
-static uart_rx_cb_t _rx_cb;
-static void * _cb_arg;
+static uart_rx_cb_t _rx_cb[UART_NUMOF];
+static void * _cb_arg[UART_NUMOF];
 
-void UART0_IRQHandler(void) __attribute__((interrupt("IRQ")));
+typedef void (*irq_fun_t)(void);
+
+#if UART_NUMOF > 0
+static void UART0_IRQHandler(void) __attribute__((interrupt("IRQ")));
+#endif
+#if UART_NUMOF > 1
+static void UART1_IRQHandler(void) __attribute__((interrupt("IRQ")));
+#endif
+#if UART_NUMOF > 2
+static void UART2_IRQHandler(void) __attribute__((interrupt("IRQ")));
+#endif
+#if UART_NUMOF > 3
+static void UART3_IRQHandler(void) __attribute__((interrupt("IRQ")));
+#endif
+
+/**
+ * @brief   Get the pointer to the base register of the given UART device
+ *
+ * @param[in] dev       UART device identifier
+ *
+ * @return              base register address
+ */
+static inline lpc23xx_uart_t *get_dev(uart_t dev)
+{
+    return uart_config[dev].dev;
+}
+
+/* get the UART number from the address */
+static inline uint8_t _uart_num(lpc23xx_uart_t* uart)
+{
+    switch ((uint32_t) uart) {
+    case UART0_BASE_ADDR:
+        return 0;
+    case UART1_BASE_ADDR:
+        return 1;
+    case UART2_BASE_ADDR:
+        return 2;
+    case UART3_BASE_ADDR:
+        return 3;
+    }
+
+    return 0;
+}
+
+static const uint8_t _uart_int[] = {
+    UART0_INT, UART1_INT, UART2_INT, UART3_INT
+};
+
+static irq_fun_t _uart_isr[UART_NUMOF] = {
+#if UART_NUMOF > 0
+    UART0_IRQHandler,
+#endif
+#if UART_NUMOF > 1
+    UART1_IRQHandler,
+#endif
+#if UART_NUMOF > 2
+    UART2_IRQHandler,
+#endif
+#if UART_NUMOF > 3
+    UART3_IRQHandler,
+#endif
+};
+
+/* Table for FDR register contents
+ * bits 0…3: DIVADDVAL
+ * bits 4…7: MULVAL
+ */
+#define DIV_MUL(m, d) ((d << 4) | m)
```

Comment:
Still, this looks a bit misleading, as `m` divides and `d` multiplies.

Reply:
fixed.